### Hacker News Data Analysis

For this project, we are going to doing some analysis on posts and data from the Hacker News website. Hacker News is a technology site thatt features posts that with tech realed contents that get posted based on comments and upvote counts. See this link for more on the website
[link](https://thehackernews.com/)


Posts on the Hacker News website are usually classified into two groups, AskHN and Show HN. For this project we are going to try to answer and retrieve the following information:
* Do Ask HN or Show HN retrieve more comments on the average
* Do posts created at a certan time recieve ore comments on average?

#### 1. Importing the libraries and readng the data set into a list of lists

In [2]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn[0:5] 

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [3]:
headers = hn[0]
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [4]:
hn = hn[1:]
hn[1:6]

[['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12'],
 ['10482257',
  'Title II kills investment? Comcast and other ISPs are now spending more',
  'http://arstechnica.com/business/2015/10/comcast-and-other-isps-boost-network-

#### 2. Filter through the data to seperate posts beginning with Ask HN and Show HN

In [5]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(title)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))        
    
        
        

1744
1162
17194


Now let us determine which posts recieve more comments on average

In [6]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = row[4]
    total_ask_comments += int(num_comments)
    
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

total_show_comments = 0

for row in show_posts:
    num_comments2 = row[4]
    total_show_comments += int(num_comments2)
    
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)
    

14.038417431192661
10.31669535283993


Now that we have establlished that the 'ask posts' have more comments on the average, we would concentrate our next few analysis on them. Next we would try to determine any relationship between the time these 'ask posts' are created, and the number of comments generated.

We are going to do this in two stages. First, we would:
1. Calculate the number of posts created in each hour of the day with the correxponding number of comments
2. Calculate the average number of askposts recieved per hour created.

#### 1. Number of ask posts and corresponing number of comments per hour

In [ ]:
import datetime as dt

result_list = [] #this will be a list of lists
for row in ask_posts:
    result_list.append([row[6], int(row[4])])
    
comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour
counts_by_hour

    


Now that we have created two dictionaries, one with the number of ask posts per hour while the other is the number of comments of these posts each hour. Next, we would use these two data and calculate the average number of comments per post each hour. 


#### 2. Calculate the average comments for each post for each hour.

In [10]:
avg_by_hour = []

avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['09', 5.5777777777777775],
 ['12', 9.41095890410959],
 ['13', 14.741176470588234],
 ['05', 10.08695652173913],
 ['20', 21.525],
 ['00', 8.127272727272727],
 ['22', 6.746478873239437],
 ['01', 11.383333333333333],
 ['11', 11.051724137931034],
 ['19', 10.8],
 ['21', 16.009174311926607],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['02', 23.810344827586206],
 ['17', 11.46],
 ['03', 7.796296296296297],
 ['10', 13.440677966101696],
 ['18', 13.20183486238532],
 ['15', 38.5948275862069],
 ['08', 10.25]]

From the output above we can see that it is difficult the see the times with more comments. It would be a great idea to sort the list.

In [12]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap
    

[[5.5777777777777775, '09'], [9.41095890410959, '12'], [14.741176470588234, '13'], [10.08695652173913, '05'], [21.525, '20'], [8.127272727272727, '00'], [6.746478873239437, '22'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [16.009174311926607, '21'], [7.985294117647059, '23'], [7.852941176470588, '07'], [9.022727272727273, '06'], [7.170212765957447, '04'], [13.233644859813085, '14'], [16.796296296296298, '16'], [23.810344827586206, '02'], [11.46, '17'], [7.796296296296297, '03'], [13.440677966101696, '10'], [13.20183486238532, '18'], [38.5948275862069, '15'], [10.25, '08']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [13]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.



### Conclusion

From our analysis we can say that posts that generate more engagements in terms of comments are usually those posted by around 15:00 or 3pm EST. 